src/cockpit/channels/filesystem.py

```diff
@@ -77,39 +79,35 @@ def do_open(self, options):
             self.close()
 
 
-class FsReadChannel(Channel):
+class FsReadChannel(GeneratorChannel):
     payload = 'fsread1'
 
-    def do_open(self, options):
-        self.ready()
+    def get_data(self, size_hint: int, options: Dict[str, object]) -> Generator[bytes, int, Dict[str, object]]:
+        binary = options.get('binary', False)
+        max_read_size = options.get('max_read_size')
+
+        logger.debug('Opening file "%s" for reading', options['path'])
+
         try:
-            logger.debug('Opening file "%s" for reading', options['path'])
-            try:
-                with open(options['path'], 'rb') as filep:
-                    buf = os.stat(filep.fileno())
-                    tag = tag_from_stat(buf)
-                    if max_read_size := options.get('max_read_size'):
-                        if buf.st_size > max_read_size:
-                            raise ChannelError('too-large')
-
-                    data = filep.read()
-            except FileNotFoundError:
-                self.close(tag='-')
-                return
-            except PermissionError:
-                raise ChannelError('access-denied')
-            except OSError:
-                raise ChannelError('internal-error')
+            with open(options['path'], 'rb') as filep:
+                buf = os.stat(filep.fileno())
+                if max_read_size is not None and buf.st_size > max_read_size:
+                    raise ChannelError('too-large')
 
-            if 'binary' not in options:
-                data = data.replace(b'\0', b'').decode('utf-8', errors='ignore').encode('utf-8')
+                while (data := filep.read(size_hint)):
```

Comment:
Let's please avoid walruses. We still have a few, but let's not increase that problem.

Reply:
Also, can we make that a `.read1()`. I don't think it makes much sense to read multiple chunks without sending  them out in between. cockpit.file() can read pipes, files from NFS, etc.

Reply:
OK.